src/Features/Core/Portable/EmbeddedLanguages/Json/LanguageServices/JsonEmbeddedClassifier.cs

```diff
@@ -0,0 +1,206 @@
+
+        public override ImmutableArray<int> SyntaxTokenKinds { get; }
+
+        public JsonEmbeddedClassifier(EmbeddedLanguageInfo info)
+        {
+            _info = info;
+            SyntaxTokenKinds = ImmutableArray.Create(info.StringLiteralTokenKind);
+        }
+
+        public override void AddClassifications(
+            SyntaxToken token,
+            SemanticModel semanticModel,
+            ClassificationOptions options,
+            ArrayBuilder<ClassifiedSpan> result,
+            CancellationToken cancellationToken)
+        {
+            if (_info.StringLiteralTokenKind != token.RawKind)
+                return;
+
+            if (!options.ColorizeJsonPatterns)
+                return;
+
+            var detector = JsonLanguageDetector.GetOrCreate(semanticModel.Compilation, _info);
+            var tree = detector.TryParseString(token, semanticModel, cancellationToken);
+            if (tree == null)
+                return;
+
+            var visitor = s_visitorPool.Allocate();
+            try
+            {
+                visitor.Result = result;
+                AddClassifications(tree.Root, visitor, result);
+            }
+            finally
+            {
+                visitor.Result = null;
+                s_visitorPool.Free(visitor);
+            }
+        }
+
+        private static void AddClassifications(JsonNode node, Visitor visitor, ArrayBuilder<ClassifiedSpan> result)
+        {
+            node.Accept(visitor);
+
+            foreach (var child in node)
+            {
+                if (child.IsNode)
+                {
+                    AddClassifications(child.Node, visitor, result);
+                }
+                else
+                {
+                    AddTriviaClassifications(child.Token, result);
+                }
+            }
+        }
+
+        private static void AddTriviaClassifications(JsonToken token, ArrayBuilder<ClassifiedSpan> result)
+        {
+            foreach (var trivia in token.LeadingTrivia)
+                AddTriviaClassifications(trivia, result);
+
+            foreach (var trivia in token.TrailingTrivia)
+                AddTriviaClassifications(trivia, result);
+        }
+
+        private static void AddTriviaClassifications(JsonTrivia trivia, ArrayBuilder<ClassifiedSpan> result)
+        {
+            if (trivia.Kind is JsonKind.MultiLineCommentTrivia or JsonKind.SingleLineCommentTrivia &&
+                trivia.VirtualChars.Length > 0)
+            {
+                result.Add(new ClassifiedSpan(
+                    ClassificationTypeNames.JsonComment, GetSpan(trivia.VirtualChars)));
+            }
+        }
+
+        private class Visitor : IJsonNodeVisitor
+        {
+            public ArrayBuilder<ClassifiedSpan>? Result;
+
+            private void AddClassification(JsonToken token, string typeName)
+            {
+                if (!token.IsMissing)
+                    Result!.Add(new ClassifiedSpan(typeName, token.GetSpan()));
```

Comment:
Not sure if you want to keep Result non-nullable and suppress the null assignment when we move it back to the pool -- that way this code is annotated for the "correct" use and evil is in one limited place.

Reply:
that works!

Reply:
actually, it's just as bad.  you get a warning on assigning to null, and that the constructor doesn't assign it...